scylla-proxy/src/proxy.rs

```diff
@@ -0,0 +1,766 @@
+                        return Err(DoorkeeperError::NoMorePorts);
+                    }
+                    desired_addr.set_port(next_port);
+                }
+                desired_addr
+            };
+
+            socket.connect(self.node.real_addr).await.map(|ok| {
+                info!(
+                    "Connected to the cluster from {} at {}, shard {}.",
+                    ok.local_addr().unwrap(),
+                    self.node.real_addr,
+                    shard_preserving_addr.port() % shards
+                );
+                ok
+            })
+        } else {
+            TcpStream::connect(self.node.real_addr).await.map(|ok| {
+                info!("Connected to the cluster at {}.", self.node.real_addr);
+                ok
+            })
+        }
+        .map_err(|err| DoorkeeperError::NodeConnectionAttempt(self.node.real_addr, err))?;
+
+        Ok((driver_addr, driver_stream, cluster_stream))
+    }
+
+    fn next_port_to_same_shard(&self, port: u16) -> u16 {
+        port.wrapping_add(self.shards_number.unwrap())
+    }
+
+    async fn obtain_shards_number(
+        &self,
+        mut connection: TcpStream,
+    ) -> Result<u16, DoorkeeperError> {
+        write_frame(
+            HARDCODED_OPTIONS_PARAMS,
+            FrameOpcode::Request(RequestOpcode::Options),
+            &Bytes::new(),
+            &mut connection,
+        )
+        .await
+        .map_err(DoorkeeperError::ObtainingShardNumber)?;
+
+        let supported_frame = read_response_frame(&mut connection)
+            .await
+            .map_err(DoorkeeperError::ObtainingShardNumberFrame)?;
+
+        let options = read_string_multimap(&mut supported_frame.body.as_ref())
+            .map_err(DoorkeeperError::ObtainingShardNumberParseOptions)?;
+        let nr_shards_entry = options.get("SCYLLA_NR_SHARDS");
+        let shards = match nr_shards_entry
+            .and_then(|vec| vec.first())
+            .ok_or(DoorkeeperError::ObtainingShardNumberNoShardInfo)?
+            .parse::<u16>()
+            .map_err(DoorkeeperError::ObtainingShardNumberParseShardNumber)?
+        {
+            0u16 => Err(DoorkeeperError::ObtainingShardNumberGotZero),
+            num => Ok(num),
+        }?;
+        info!(
+            "Obtained shards number on node {}: {}",
+            self.node.real_addr, shards
+        );
+        Ok(shards)
+    }
+}
+
+struct ProxyWorker {
+    terminate_notifier: TerminateNotifier,
+    finish_guard: FinishGuard,
+    connection_close_notifier: ConnectionCloseNotifier,
+    driver_addr: SocketAddr,
+    real_addr: SocketAddr,
+    proxy_addr: SocketAddr,
+}
+
+impl ProxyWorker {
+    fn exit(self, duty: &'static str) {
+        debug!(
+            "Worker exits: [driver: {}, proxy: {}, node: {}]::{}.",
+            self.driver_addr, self.proxy_addr, self.real_addr, duty
+        );
+        std::mem::drop(self.finish_guard);
+    }
+
+    async fn receiver_from_driver(
```

Comment:
Each of the workers have some boilerplate related to handling termination and responding with an error, I think it could be abstracted away... I tried doing it with one of the functions and it seems possible:

```rust
async fn run_until_interrupted<F, Fut>(mut self, worker_name: &'static str, f: F)
where
    F: FnOnce(SocketAddr, SocketAddr) -> Fut,
    Fut: Future<Output = Result<(), ProxyError>>,
{
    let fut = f(self.driver_addr, self.proxy_addr);

    tokio::select! {
        result = fut => {
            if let Err(err) = result {
                // error_propagator could be a field
                let _ = self.error_propagator.send(err);
            }
        }
        _ = self.terminate_notifier.recv() => {},
        _ = self.connection_close_notifier.recv() => {},
    }
    self.exit(worker_name);
}

async fn receiver_from_driver(
    self,
    mut read_half: (impl AsyncRead + Unpin),
    request_processor_tx: mpsc::UnboundedSender<RequestFrame>,
) {
    self.run_until_interrupted(
        "receiver_from_driver",
        |driver_addr, proxy_addr| async move {
            loop {
                let frame = frame::read_request_frame(&mut read_half)
                    .await
                    .map_err(|err| {
                        warn!("Request reception from {} error: {}", driver_addr, err);
                        WorkerError::DriverDisconnected(driver_addr)
                    })?;

                debug!(
                    "Intercepted Driver ({}) -> Cluster ({}) frame. opcode: {:?}.",
                    driver_addr, proxy_addr, &frame.opcode
                );
                if request_processor_tx.send(frame).is_err() {
                    warn!("request_processor had exited.");
                    return Result::<(), ProxyError>::Ok(());
                }
            }
        },
    )
    .await
}
```

This seems to reduce the amount of indents a bit.